[SNAP-2101] fix extra buffer copying and PK lookup on column tables

sumwale opened the pull request:
As noted in https://jira.snappydata.io/browse/SNAP-2101, this fixes the two issues mentioned
there and has a few cleanups. One is elimination of extra direct buffer copying to bytes and
back, and second is making the column store table PK columns use a LocalHash1Index instead
of GlobalHash so that it can be used effectively by the UNION ALL query fired by connector.

## Changes proposed in this pull request

- cleaned up the buffer ownership and copying semantics in ClientBlobs and BlobChunks:
  - now BlobChunks always own their buffer inside which should be cleared with "free()" when
    done by upper layers; ClientBlob will also own the BlobChunk inside until free()d
  - when ClientBlob.getAsLastChunk() is invoked, then it will immediately give up the chunk
    in every case and make its current chunk reference as null (i.e. blob is not usable anymore)
  - avoid Blob conversion to bytes for BufferedBlobs in SQLBinary, SQLBlob and EmbedResultSet
- added a partitioningColumns method to InternalPartitionResolver which is used in
  sqlgrammar.jj  for object tables to set the partitioning columns and policy on the
  DistributionDefinitionNode; this gets used in CreateTableConstantAction to create Hash1Index
  instead of GlobalHash for PK
- add "425" prefix to authorization related failures rather than throwing as SQLSyntaxError
- increased parking time on zero async writes to a more reasonable 500ns
- made KEEPALIVE_INTVL to less aggressive 4s by default particularly for remote clients
- allow for "snappydata." prefix for SystemProperties(), so snappydata need not use "gemfirexd."
- correct many cases of ambiguous grammar (also eliminates most warnings in javacc compile):
  - make IF as reserved keyword to resolve ambiguities in "DROP IF EXISTS ..." kind of DDLs
  - use integer/long literal non-terminals for simpleValueSpeficiation_* calls instead of
    accepting arbitrary strings and then casting to integer/long
- moved ClientSharedUtils.equalsBuffers to inside ConnectionHolder since it makes assumptions
  about the byte buffers (that left one is always a heap buffer)
- use a common ClientSharedData.NULL_BUFFER for empty heap buffer

## Patch testing

precheckin -Pstore, manual performance runs of ColumnCacheBenchmark in connector mode

## ReleaseNotes changes

NA

## Other PRs 

https://github.com/SnappyDataInc/snappydata/pull/895

Commit:
[SNAP-2101] fix extra buffer copying and PK lookup on column tables (by columnID)

- cleaned up the buffer ownership and copying semantics in ClientBlobs and BlobChunks:
  - now BlobChunks always own their buffer inside which should be cleared with "free()" when done
    by upper layers; ClientBlob will also own the BlobChunk inside
  - when ClientBlob.getAsLastChunk is invoked, then it will immediately give up the chunk
    in every case and transfer making its inner chunk as null (i.e. blob is not usable anymore)
  - avoid Blob conversion to bytes for BufferedBlobs in SQLBinary, SQLBlob and EmbedResultSet methods
- added a partitioningColumns method to InternalPartitionResolver which is used in sqlgrammar.jj
  for object tables to set the partitioning columns and policy on the DistributionDefinitionNode;
  this gets used in CreateTableConstantAction to create Hash1Index instead of GlobalHash for PK
- add "425" to auth related failures rather than throwing as SQLSyntaxError
- increased parking time on zero async writes to a more reasonable 500ns
- made KEEPALIVE_INTVL to less aggressive 4s by default particularly for remote client connections
- allow for "snappydata." prefix for SystemProperties(), so snappydata need not use "gemfirexd."
- correct many cases of ambiguous grammar (also eliminates most warnings in javacc compilation):
  - make IF as reserved keyword to resolve ambiguities in "DROP IF EXISTS ..." kind of DDLs
  - use integer/long literal non-terminals for simpleValueSpeficiation_* calls instead of
    accepting arbitrary strings and then casting to integer/long
- moved ClientSharedUtils.equalsBuffers to inside ConnectionHolder since it makes assumptions
  about the byte buffers (that left one is always a heap buffer)
- use a common ClientSharedData.NULL_BUFFER for empty heap buffer

Commit:
formatting

Commit:
GEODE-2240: unsafe concurrent access in expiredTombstones

Commit:
Fixing queue handling in TombstoneService

- Removed queue.size() check which is expensive for linked queues
  (causes CPU to peg with simple scenarios like rollover from row buffer)
- Fixed unsafe access to currentTombstone. Double-check lock should be on volatile.

rishitesh commented:
@sumwale Apart from performance, would not it solve another problem, where we had to provision more heap memory keeping in view that embedded data were getting copied to heap space before sending ?

sumwale commented:
@rishitesh Yes for client-server. For that I have also been testing with reduced "jdk.nio.maxCachedBufferSize" to 256k because many layers like parquet reader (netty too?) use heap buffers. Came to know recently about this property that JVM caches for native transfer as thread-local. So each thread will keep holding on to the largest DirectByteBuffer it ended up creating due to use of heap buffers in channel read/write. This property will help limit it while the JDK by default has no upper limit.

Commit:
fixed NPE with "PARTITION BY PRIMARY KEY" and cleanups

- previous change to set/validate partitioning columns in parsing had a problem
  with primary key partitioning where columns are inferred; skipping that case
- refactored busy wait on zero read (from non-blocking channel)

Commit:
fix for a failure in StatementPlanDUnit

Commit:
fix occasional failure in ExpirationDUnit

Commit:
fix failure in CreateDiskStoreTest

Commit:
Merge remote-tracking branch 'origin/snappy/master' into SNAP-2101
